sqlglot/diff.py

```diff
@@ -108,44 +104,47 @@ def diff(
         expression tree into the target one.
     """
     matchings = matchings or []
-    matching_ids = {id(n) for pair in matchings for n in pair}
 
     def compute_node_mappings(
-        original: exp.Expression, copy: exp.Expression
+        old_nodes: tuple[exp.Expression, ...], new_nodes: tuple[exp.Expression, ...]
     ) -> t.Dict[int, exp.Expression]:
         node_mapping = {}
-        for old_node, new_node in zip(
-            reversed(tuple(original.walk())), reversed(tuple(copy.walk()))
-        ):
-            # We cache the hash of each new node here to speed up equality comparisons. If the input
-            # trees aren't copied, these hashes will be evicted before returning the edit script.
+        for old_node, new_node in zip(reversed(old_nodes), reversed(new_nodes)):
             new_node._hash = hash(new_node)
-
-            old_node_id = id(old_node)
-            if old_node_id in matching_ids:
-                node_mapping[old_node_id] = new_node
+            node_mapping[id(old_node)] = new_node
 
         return node_mapping
 
+    # if the source and target have any shared objects, that means there's an issue with the ast
+    # the algorithm won't work because the parent / hierarchies will be inaccurate
+    source_nodes = tuple(source.walk())
+    target_nodes = tuple(target.walk())
+
+    copy = {id(n) for n in source_nodes} & {id(n) for n in target_nodes}
     source_copy = source.copy() if copy else source
     target_copy = target.copy() if copy else target
 
-    node_mappings = {
-        **compute_node_mappings(source, source_copy),
-        **compute_node_mappings(target, target_copy),
-    }
-    matchings_copy = [(node_mappings[id(s)], node_mappings[id(t)]) for s, t in matchings]
-
-    edit_script = ChangeDistiller(**kwargs).diff(
-        source_copy,
-        target_copy,
-        matchings=matchings_copy,
-        delta_only=delta_only,
-    )
-
-    if not copy:
-        for node in chain(source.walk(), target.walk()):
-            node._hash = None
+    try:
+        # We cache the hash of each new node here to speed up equality comparisons. If the input
+        # trees aren't copied, these hashes will be evicted before returning the edit script.
+        if copy:
```

Comment:
Should this be `copy and matchings` to avoid unnecessarily calls to `compute_node_mappings`?